smartcontract/service/native/param.go

```diff
@@ -0,0 +1,204 @@
+		getAdminStorageItem(destinationAdmin))
+	return nil
+}
+
+func SetGlobalParam(native *NativeService) error {
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	getAdmin(native, contract)
+	if !native.ContextRef.CheckWitness(common.Address(*admin)) {
+		return errors.NewErr("[Set Param]Authentication failed!")
+	}
+	params := new(states.Params)
+	if err := params.Deserialize(bytes.NewBuffer(native.Input)); err != nil {
+		return errors.NewErr("[Set Param]Deserialize failed!")
+	}
+	// read old param from database
+	storageParams, err := getStorageParam(native, getParamKey(contract, PREPARE_VALUE))
+	if err != nil{
+		return err
+	}
+	// update param
+	for key, value := range *params{
+		(*storageParams)[key] = value
+	}
+	native.CloneCache.Add(scommon.ST_STORAGE, getParamKey(contract, PREPARE_VALUE),
+		getParamStorageItem(storageParams))
+	notifyParamSetSuccess(native, contract, *params)
+	return nil
+}
+
+func CreateSnapshot(native *NativeService) error {
+	contract := native.ContextRef.CurrentContext().ContractAddress
+	getAdmin(native, contract)
+	if !native.ContextRef.CheckWitness(common.Address(*admin)) {
+		return errors.NewErr("[Create Snapshot]Authentication failed!")
+	}
+	// read prepare param
+	prepareParam, err := getStorageParam(native, getParamKey(contract, PREPARE_VALUE))
+	if err != nil {
+		return errors.NewDetailErr(err, errors.ErrNoCode, "[Create Snapshot] storage error!")
+	}
+	if len(*prepareParam) == 0 {
+		return errors.NewErr("[Create Snapshot] Prepare param doesn't exist!")
+	}
+	// set prepare value to current value, make it effective
+	native.CloneCache.Add(scommon.ST_STORAGE, getParamKey(contract, CURRENT_VALUE), getParamStorageItem(prepareParam))
+	// clear memory cache
+	clearCache()
+	return nil
+}
+
+func getAdmin(native *NativeService, contract common.Address) {
+	if admin == nil || *admin == *new(states.Admin) {
+		var err error
+		// get admin from database
+		admin, err = getStorageAdmin(native, getAdminKey(contract, false))
+		// there are no admin in database
+		if err != nil {
+			initAddress := ctypes.AddressFromPubKey(account.GetBookkeepers()[0])
+			copy((*admin)[:], initAddress[:])
+		}
+	}
+}
+
+func clearCache() {
+	paramCache.lock.Lock()
+	defer paramCache.lock.Unlock()
+	paramCache.Params = make(map[string]string)
+}
+
+func setCache(params *states.Params) {
+	paramCache.lock.Lock()
+	defer paramCache.lock.Unlock()
+	paramCache.Params= *params
+}
+
+func getParamFromCache(key string) string {
+	paramCache.lock.RLock()
+	defer paramCache.lock.RUnlock()
+	return paramCache.Params[key]
+}
+
+func RegisterParamContract(native *NativeService) {
+	native.Register("init", ParamInit)
+	native.Register("acceptAdmin", AcceptAdmin)
+	native.Register("transferAdmin", TransferAdmin)
+	native.Register("setGlobalParam", SetGlobalParam)
+	native.Register("createSnapshot", CreateSnapshot)
+}
+
+func GetGlobalPram(native *NativeService, paramName string) (string, error) {
```

Comment:
fix function name typos  here and the documentation

Reply:
sorry...